zend_abstract_interface/exceptions/php8/exceptions.c

```diff
@@ -0,0 +1,99 @@
+#include "exceptions.h"
+
+#include <Zend/zend_smart_str.h>
+
+zend_string *zai_exception_message(zend_object *ex) {
+    zval *message = ZAI_EXCEPTION_PROPERTY(ex, ZEND_STR_MESSAGE);
+    if (Z_TYPE_P(message) != IS_STRING) {
+        // interned, so we can forego freeing it ourselves
+        return zend_string_init_interned(ZEND_STRL("(internal error reading exception message)"), 1);
+    }
+    return Z_STR_P(message);
+}
+
+/* Modeled after Exception::getTraceAsString:
+ * @see https://heap.space/xref/PHP-8.0/Zend/zend_exceptions.c#getTraceAsString
+ */
+zend_string *zai_get_trace_without_args(zend_array *trace) {
+    zval *frame;
+    smart_str str = {0};
+    uint32_t num = 0;
+    ZEND_HASH_FOREACH_VAL(trace, frame) {
+        smart_str_appendc(&str, '#');
+        smart_str_append_long(&str, num++);
+        smart_str_appendc(&str, ' ');
+
+        if (UNEXPECTED(Z_TYPE_P(frame) != IS_ARRAY)) {
+            smart_str_appends(&str, "[invalid frame]\n");
+            continue;
+        }
+
+        zend_array *ht = Z_ARRVAL_P(frame);
+
+        zval *file = zend_hash_find_ex(ht, ZSTR_KNOWN(ZEND_STR_FILE), 1);
+        if (file) {
+            if (Z_TYPE_P(file) != IS_STRING) {
+                /* This is not the function, but the file. However, this is what
+                 * PHP itself reports, so we'll match it for consistency.
+                 */
+                smart_str_appends(&str, "[unknown function]");
```

Comment:
In 8.0.7 this is fixed, btw: https://github.com/php/php-src/commit/6fd13d0afd8f487713fa068aec9fc83358319a48.

Reply:
Good point - I will change that :-)